### PR TITLE
Make state rebuilding more deterministic

### PR DESCRIPTION
**Pull request by hevrard**
This is a big but boring change to reduce the sources of non-determinism in the state rebuilder. It is only a first step, but it already massively reduces differences in the list of generated commands.

Make sure to review commit-by-commit, to have piecemeal diffs.

Tested on an MEC trace that generates 50k+ commands for state rebuilding, these changes brings the number of diff chunks (between the lists of state-rebuilding commands generated by several calls to `linearize_trace`) from ~2000 down to ~30.

Bug: b/155048196

### CODE REVIEW

File: gapis/api/vulkan/state_rebuilder.go [hunk 1]
```diff
@@ -461,11 +461,13 @@ func (sb *stateBuilder) write(cmd api.Cmd) {
 
 func (sb *stateBuilder) createInstance(vk VkInstance, inst InstanceObjectʳ) {
 	enabledLayers := []Charᶜᵖ{}
-	for _, layer := range inst.EnabledLayers().All() {
+	for _, k := range inst.EnabledLayers().Keys() {
+		layer := inst.EnabledLayers().Get(k)
```
Comment (chrisforbes) [inline gapis/api/vulkan/state_rebuilder.go:465]: Whoa -- we must preserve the original order here! Layer order matters.

File: gapis/api/vulkan/state_rebuilder.go [hunk 2]
```diff
@@ -681,18 +683,21 @@ func (sb *stateBuilder) createSurface(s SurfaceObjectʳ) {
 
 func (sb *stateBuilder) createDevice(d DeviceObjectʳ) {
 	enabledLayers := []Charᶜᵖ{}
-	for _, layer := range d.EnabledLayers().All() {
+	for _, k := range d.EnabledLayers().Keys() {
```
Comment (chrisforbes) [inline gapis/api/vulkan/state_rebuilder.go:686]: ditto

File: gapis/api/vulkan/state_rebuilder.go [hunk 3]
```diff
@@ -681,18 +683,21 @@ func (sb *stateBuilder) createSurface(s SurfaceObjectʳ) {
...
 		enabledLayers = append(enabledLayers, NewCharᶜᵖ(sb.MustAllocReadData(layer).Ptr()))
 	}
 	enabledExtensions := []Charᶜᵖ{}
-	for _, ext := range d.EnabledExtensions().All() {
+	for _, k := range d.EnabledExtensions().Keys() {
+		ext := d.EnabledExtensions().Get(k)
 		enabledExtensions = append(enabledExtensions, NewCharᶜᵖ(sb.MustAllocReadData(ext).Ptr()))
 	}
 
 	queueCreate := map[uint32]VkDeviceQueueCreateInfo{}
```
Comment (chrisforbes) [inline gapis/api/vulkan/state_rebuilder.go:696]: You're still iterating over this map on :724 below..

File: gapis/api/vulkan/state_rebuilder.go (hunk 1 above)
Comment (hevrard) [reply gapis/api/vulkan/state_rebuilder.go:465]: I think this change actually fixes it, as the EnabledLayers here is a map of index to layer names, and they are stored in order during capture:

https://github.com/google/agi/blob/master/gapis/api/vulkan/api/instance.api#L120-L134

So following the order of this map keys (as we do here with ranging on `Keys()`) should respect the original order.

File: gapis/api/vulkan/state_rebuilder.go (hunk 3 above)
Comment (hevrard) [reply gapis/api/vulkan/state_rebuilder.go:696]: Fixed.

File: gapis/api/vulkan/state_rebuilder.go (hunk 1 above)
Comment (chrisforbes) [reply gapis/api/vulkan/state_rebuilder.go:465]: :+1:

**Pull request merged**